rustls/src/time_provider.rs

```diff
@@ -0,0 +1,50 @@
+//! no-std time provider
+
+use alloc::sync::Arc;
+use core::fmt::Debug;
+
+use pki_types::UnixTime;
+
+/// An object that provides the current time.
+///
+/// This is used to, for example, check if a certificate has expired during certificate validation,
+/// or to check the age of a ticket.
+#[derive(Clone, Debug)]
+pub struct TimeProvider {
+    inner: Arc<dyn GetCurrentTime>,
+}
+
+impl TimeProvider {
+    /// Creates a new time provider
+    pub fn new(time_getter: impl GetCurrentTime + 'static) -> Self {
+        Self {
+            inner: Arc::new(time_getter),
+        }
+    }
+
+    /// A time provider that always fails
+    pub fn none() -> Self {
```

Comment:
How does it make sense to have this? Is it possible to do TLS without time?

Reply:
As mentioned in [the RFC](https://github.com/japaric/rustls/blob/ja-no-std-support-rfc/rfcs/000-no-std-support.md#systemtimenow) this is the default time provider and will always error as the user must provide their own time provider in no-std contexts.

Reply:
But why do we want to provide a default time provider that doesn't actually provide time? Is it load-bearing that time providers can yield `None`?

Reply:
So as far as I can surmise, the only reason to have `TimeProvider::current_time()` yield `Option` is that the config builder must create a `TImeProvider` to build a complete config, and in `no_std` situation that is platform dependent so we can't create it. I think we should probably force `no_std` configurations through a new separate `ClientConfig::builder_with_time_provider()` path instead and make `TimeProvider` yield just `UnixTime` or, more likely, `Result<UnixTime, Error>`.

Reply:
some real-time clock drivers are fallible [ds1307](https://docs.rs/ds1307/latest/ds1307/#set-and-get-the-current-date-and-time-all-at-once) is an example of this. Returning `Option<UnixTime>` allows representing such errors, you could argue that we could return `Result<UnixTime, TimeProvider::Error>` or something like that but I consider that the error provides little to no use.